File: app/src/main/java/com/digis/android/task/data/source/remote/Response.java
package com.digis.android.task.data.source.remote;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public final class Response<D> {
    private D data;
    private String message;
    private boolean isSuccessful;

    private Response() { }

    public static <T> Response<T> create(@NonNull HttpResponse<T> httpResponse) {
        Response<T> response = new Response<>();
        response.data = httpResponse.getData();
        response.isSuccessful = (httpResponse.getData() != null);
        return response;
    }

    public static <T> Response<T> create(@NonNull Application app, @NonNull Throwable throwable,
                                         @Nullable T data) {
        Response<T> response = new Response<>();
        response.data = data;
        response.message = ErrorHandler.getMessage(app, throwable);
        response.isSuccessful = false;
        return response;
    }

    public D getData() {
        return data;
    }

    @Nullable
    public String getMessage() {
        return message;
    }

    public boolean isSuccessful() {
        return isSuccessful;
    }
}

File: app/src/main/java/com/digis/android/task/implementation/kotlin/data/source/remote/Response.kt
package com.digis.android.task.implementation.kotlin.data.source.remote

import android.app.Application

sealed class Response<out D> {
    data class Success<out T>(val data: T) : Response<T>()
    data class Error(val message: String?) : Response<Nothing>()

    companion object {
        @JvmStatic
        fun <T> create(httpResponse: HttpResponse<T>): Response<T> {
            return if (httpResponse.isSuccessful) {
                Success(data = httpResponse.data!!)
            } else {
                Error(message = null)
            }
        }

        @JvmStatic
        fun <T> create(app: Application, throwable: Throwable): Response<T> =
                Error(message = ErrorHandler.getMessage(app, throwable))

        private val <T> HttpResponse<T>.isSuccessful get() = (code in 200..299) && data != null
    }
}